update eyre documentation

tinnus-napbus opened the pull request:
removes old documentation and adds a scry ref, data types ref, external
& internal api ref, and examples.

@drbeefsupreme

tinnus-napbus commented:
note this requires the change to faq.md in https://github.com/urbit/urbit.org/pull/642 in order for zola to build it successfully

Fang- reviewed (changes requested):
Wow! Another big-boy PR right here. Very thorough, complete and clear documentation. Thank you for this!

The only thing I'm really missing here is some description of the header handling that Eyre does for you. CORS headers come up implicitly, but the behavior isn't talked about in too much detail (and extends to `OPTIONS` requests wholesale). Eyre respects the `Forwarded` header when figuring out the `address` for inbound requests (but only if the request came from localhost). Additionally, Eyre (and/or the runtime?) take care of the content-length header for you.

I also left some corrections, and a bunch of suggestions and style nits in the comments below. Hope they're useful, and let me know if you want me to further clarify on anything here. (:

timlucmiptev commented:
@tinnus-napbus fang had some changes requested here, wanted me to re-ping you

tinnus-napbus commented on arvo/eyre/external-api-ref.md:
```diff
@@ -0,0 +1,399 @@
+
+### Ack
+
+This is for acknowledging a [fact](#fact) from a Gall agent to which you've subscribed.
```
the only other possible responses are a kick, poke-ack and watch-ack, and afaict none of those require an ack, only a `diff` type response which is only triggered by a fact (and they can only come from agents cos you can't poke or watch vanes)

```
          %fact
        :~  ['response' [%s 'diff']]
          ::
            :-  'json'
            ~|  [%unexpected-fact-mark p.cage.sign]
            ?>  =(%json p.cage.sign)
            ;;(json q.q.cage.sign)
        ==
```

tinnus-napbus replied:
hmm no perhaps you're right, it seems it wants them for all. I was confused I think cos it's weird to ack an ack

Fang- replied:
You aren't acking the `poke-ack` itself, you are acking the SSE message that notified you of the `poke-ack`.

This is more obvious in the broader context: Eyre keeps a queue of message it has sent over the channel, because it doesn't know whether you're still connected or not. (And might want to send you the ones you missed when you reconnect.) Acking a message lets Eyre remove it from the queue, because you've confirmed you did indeed receive it.

tinnus-napbus commented on arvo/eyre/tasks.md:
````diff
@@ -0,0 +1,243 @@
+
+This `task` is how Eyre receives an inbound HTTP request over the local loopback port. It behaves the same and takes the same arguments as in the [%request](#request) example except it skips any normally required authentication. Just like for a [%request](#request) `task`, you'd not normally use this manually.
+
+# `%cancel-request`
+
+This `task` is sent to Eyre by Vere when an HTTP client closes its previously established connection. You would not normally use this manually.
+
+## Accepts
+
+```hoon
+~
+```
+
+Takes no arguments.
+
+## Returns
````
I agree, I initially wrote these along with the clay docs that way, but I changed it to bring it inline with the format of the existing docs. Probably should change all of them. @drbeefsupreme what do you think

drbeefsupreme replied:
Yeah I agree, it does look better to just give the noun format given in `lull` directly rather than separating out the arguments. Especially when, someday, the code in the docs is pulled directly from the codebase. So I agree with these change, and we should go back and change all the other API docs (sorry for making you change the Clay docs!) to look like this.

Fang- commented on arvo/eyre/examples.md:
```diff
@@ -0,0 +1,689 @@
+
+Here we'll look at approving and rejecting a CORS origin by passing Clay a [%approve-origin](@/docs/arvo/eyre/tasks.md#approve-origin) `task` and [%reject-origin](@/docs/arvo/eyre/tasks.md#reject-origin) `task` respectively.
+
+First, using `|pass` in the dojo, let's approve the origin `http://foo.example` by sending Eyre a `%approve-origin` `task`:
```
Any reason for sticking with `|pass` here instead of the manual task & scry construction? If it's for educational purposes, we should probably at least mention the generators as the thing you'll want for every-day use.

Fang- commented on arvo/eyre/data-types.md:
````diff
@@ -0,0 +1,290 @@
+  ==
+```
+
+An inbound HTTP request and metadata - this is what is poked into a bound app. The `authenticated` field says whether the request was made with a valid session cookie. The `secure` field says whether it was made with HTTPS. The [$address](#address) is the client's IP address. The [$request:http](#request-http) contains the HTTP request itself. 
````
Here and elsewhere, I'm not sure how common "poke into" is in the rest of the docs. It feels a little off to me, compared to the slightly more verbose "app is poked with ...". No strong feelings, but want to offer up this alternative for your consideration.

Fang- commented on arvo/eyre/external-api-ref.md:
```diff
@@ -0,0 +1,399 @@
+
+### Ack
+
+This is for acknowledging an SSE event. If you `ack` one event, you also implicitly `ack` all previous events. Events must be `ack`ed.
```
Either here or in the examples (or both?) might be worth mentioning explicitly that they must be acked _so that Eyre can forget them_, and that if you don't, Eyre will eventually kill your subscription. This "clogging" behavior is mentioned briefly in the data-types doc, but should probably be elaborated somewhere.

Fang- reviewed (commented):
All good changes so far, thank you!

tinnus-napbus commented:
ok I've addressed most of the comments. The original point about covering Eyre's CORS capabilities I've DM'd @Fang- to discuss what is necessary.

On the point about spider, yeah it's technically true that the spider interface isn't directly a function of Eyre. It is kinda a de facto thing though, it's automatically bound and threads are kinda fundamental despite spider not being a vane itself. My feeling is that practically it's useful to have all the documentation of web interfaces in one place and the previous "using eyre" documentation covered threads. I probably could find a place for it in the threads documentation though, idk. What do you think @drbeefsupreme ?

On the point about the explanations of Eyre's behaviour wrt generators bound with %serve and agents bound with %connect, maybe it's not ideal to have that info on the tasks page but idk it's useful and I'm not sure where else to put it. I think putting it all in the examples page isn't the best, nor on overview because it's too technical. I could create another document that covers the %serve and %connect cases in more detail but then I'm not sure there's enough info to warrant a separate page. What do you think @drbeefsupreme ?

tinnus-napbus commented:
also sorry for not addressing this stuff earlier, I missed github's email about it and assumed nobody had got around to reviewing stuff

Fang- commented on arvo/eyre/examples.md:
```diff
@@ -526,9 +521,9 @@ The sample of the second nested gate must be:
+The return type of the generator must be [simple-payload:http](@/docs/arvo/eyre/data-types.md#simple-payload-http). If you look at our example generator you'll see it meets these requirements.
 
-If you look at our example generator you'll see it meets these requirements.
+Because generators return the entire HTTP message as a single `simple-payload`, Eyre can calculate the `content-length` itself and automatically add the header, so we don't need to do it by hand like with a Gall agent.
```
fwiw, I don't think you _need_ to specify the `content-length` header explicitly anywhere. If you don't include it, the runtime will send with `transfer-encoding: chunked` instead. Not sure if this is worth including here though.

Fang- reviewed (approved):
Thank you for all your work and changes here! This looks good to go, once @drbeefsupreme has settled the remaining discussion here and given his blessing.

> On the point about spider, yeah it's technically true that the spider interface isn't directly a function of Eyre. It is kinda a de facto thing though, it's automatically bound

Counterpoint: it's only automatically bound if the spider app is started, and many apps bind endpoints when they start, and many apps are started by default. We're not talking about file-server here either, even though that's arguably even more fundamental (functionality-wise) than spider.

drbeefsupreme commented on arvo/eyre/examples.md:
````diff
@@ -0,0 +1,624 @@
+[%pass /eyre %arvo %e %connect `/'foo' %eyre-agent]~
+```
+
+...and when `%eyre` responds with a `%bound` gift, the `+on-agent` arm will print whether the bind succeeded:
````
```suggestion
...and when `%eyre` responds with a `%bound` `gift`, the `+on-agent` arm will print whether the bind succeeded:
```

drbeefsupreme commented on arvo/eyre/examples.md:
````diff
@@ -0,0 +1,624 @@
+[unlinked from [p=~zod q=%eyre-agent]]
+```
+
+Now, first we need to bind a url to our app. In the `++  on-poke` arm, our agent will send a [%connect](@/docs/arvo/eyre/tasks.md#connect) `task` to Eyre when poked with `%bind`:
````
```suggestion
Now, first we need to bind a URL to our app. In the `++  on-poke` arm, our agent will send a [%connect](@/docs/arvo/eyre/tasks.md#connect) `task` to Eyre when poked with `%bind`:
```

drbeefsupreme commented on arvo/eyre/external-api-ref.md:
```diff
@@ -0,0 +1,379 @@
+
+If you're working with Javascript in the browser context you'll handle these with an EventSource object or by using fetch and ReadableStream. If you're using another language, there'll likely be a library available to handle SSEs.
+
+All the events that Eyre sends you on a channel must be [ack](#ack)ed so that Eyre can forget about them and clear them from the channel state. If `fact`s (as [diff](#diff)s) from Gall agents to which you've subscribed are left un`ack`ed long enough, Eyre will consider the particular subscription clogged and automatically unsubscribe you. Note that `Ack`ing one event will implicitly `ack` all pre ...
```
```suggestion
All the events that Eyre sends you on a channel must be [ack](#ack)ed so that Eyre can forget about them and clear them from the channel state. If `fact`s (as [diff](#diff)s) from Gall agents to which you've subscribed are left un`ack`ed long enough, Eyre will consider the particular subscription clogged and automatically unsubscribe you. Note that `ack`ing one event will implicitly `ack` all previous events.
```
Not sure if the uppercase Ack was supposed to imply an HTTP header? (I don't really know how HTTP works beyond what I've read here so far!)

drbeefsupreme commented on arvo/eyre/eyre.md:
```diff
@@ -2,720 +2,60 @@
-## Commentary
+# Generators
 
+Generators, which are like hoon scripts, can also be used through Eyre. Rather than having a predefined JSON API, they instead handle HTTP requests and return HTTP responses directly, and are therefore a more complex case that you're less likely to use.
```
```suggestion
Generators, which are like Hoon scripts, can also be used through Eyre. Rather than having a predefined JSON API, they instead handle HTTP requests and return HTTP responses directly, and are therefore a more complex case that you're less likely to use.
```

drbeefsupreme commented on arvo/eyre/eyre.md:
```diff
@@ -2,720 +2,60 @@
 
-An http request for `http://sampel-sipnym.urbit.org/cli` will be [redirected](dns)
-to the `%eyre` on ~sampel-sipnym, and come in as a `%this` task.
+As well as the [Channel System](#the-channel-system) and [Scries](#scrying), it's also possible for Gall Agents to deal directly with HTTP requests. This method is much more complicated that the channel system so you're unlikely to use it unless you want to build a custom HTTP-based API or something like that.
```
```suggestion
As well as the [Channel System](#the-channel-system) and [Scries](#scrying), it's also possible for Gall agents to deal directly with HTTP requests. This method is much more complicated than the channel system so you're unlikely to use it unless you want to build a custom HTTP-based API or something like that.
```

drbeefsupreme commented on arvo/eyre/eyre.md:
```diff
@@ -2,720 +2,60 @@
-which starts with a valid ship name is expected to be a full clay(well, ford)
-path, and one starting with a term implies the current serving ship and a case
-of `0`, the current revision.
+There are three generators - `+cors-registry`, `|cors-approve` and `|cors-reject` to view, approve and deny origins respectively from the dojo. Eyre also has [tasks](@/docs/arvo/eyre/tasks.md) to [approve](@/docs/arvo/eyre/tasks.md#approve-origin) and [reject](@/docs/arvo/eyre/tasks.md#reject-origin) origins programmatically, and a number of [scry endpoints](@/docs/arvo/eyre/scry.md) to query the ...
```
```suggestion
There are three generators - `+cors-registry`, `|cors-approve`, and `|cors-reject` to view, approve, and deny origins respectively from the dojo. Eyre also has [tasks](@/docs/arvo/eyre/tasks.md) to [approve](@/docs/arvo/eyre/tasks.md#approve-origin) and [reject](@/docs/arvo/eyre/tasks.md#reject-origin) origins programmatically, and a number of [scry endpoints](@/docs/arvo/eyre/scry.md) to query them. Examples are also included in the [Managing CORS Origins](@/docs/arvo/eyre/examples.md#managing-cors-origins) section of the Examples document.
```

drbeefsupreme commented on arvo/eyre/scry.md:
```diff
@@ -0,0 +1,214 @@
+
+The first few have a `care` of `x` and are a scry like `.^(<type> %ex /=//=/<some-path>)` (note the empty `desk`). The rest have no `care` and the tag replaces the `desk` like `.^(<type> %e /=<something>=)`.
+
+All examples are from the dojo.
```
```suggestion
All examples are run from the dojo.
```

drbeefsupreme commented on arvo/eyre/tasks.md:
```diff
@@ -0,0 +1,221 @@
+
+This `task` is how Eyre receives an inbound HTTP request. It will ordinarily be sent to Eyre by the runtime so you wouldn't use it except perhaps in tests.
+
+The `secure` field says whether it's over HTTPS. The `address` is the ip address from which the request originated. The [request:http](@/docs/arvo/eyre/data-types.md#request-http) is the HTTP request itself containing the method, url, headers and body.
```
```suggestion
The `secure` field says whether it's over HTTPS. The `address` is the IP address from which the request originated. The [request:http](@/docs/arvo/eyre/data-types.md#request-http) is the HTTP request itself containing the method, URL, headers, and body.
```

drbeefsupreme commented on arvo/eyre/tasks.md:
```diff
@@ -0,0 +1,221 @@
+
+## Returns
+
+Eyre may `pass` a `%response` `gift` on the appropriate `duct` depending on the contents of the `%request`, state of the connection and other factors.
```
```suggestion
Eyre may `pass` a `%response` `gift` on the appropriate `duct` depending on the contents of the `%request`, state of the connection, and other factors.
```

drbeefsupreme commented on arvo/eyre/tasks.md:
```diff
@@ -0,0 +1,221 @@
+
+# `%request-local`
+
+This `task` is how Eyre receives an inbound HTTP request over the local loopback port. It behaves the same and takes the same arguments as in the [%request](#request) example except it skips any normally required authentication. Just like for a [%request](#request) `task`, you'd not normally use this manually.
```
Need the structure and return

tinnus-napbus replied:
fixed

drbeefsupreme commented on arvo/eyre/external-api-ref.md:
```diff
@@ -0,0 +1,379 @@
+|-------------|----------------------------------------|------------|-------------------------------------------------------------------------------------|
+| `POST`      | `password=lidlut-tabwed-pillex-ridrup` | `/~/login` | The password is your web login code which can be obtained with `+code` in the dojo. |
+
+Eyre's response will include a `set-cookie` header like:
```
```suggestion
See the [examples](@/docs/arvo/eyre/examples.md#authenticating) for an example of how to do this using the curl.

Eyre's response will include a `set-cookie` header like:
```

drbeefsupreme reviewed (changes requested):
Pretty much all very minor grammar nitpicks. One of the `task` entries was also incomplete. Other than that, LGTM!

tinnus-napbus commented:
all the issues raised by @drbeefsupreme have been fixed
